File: docs/mailhub-connector.js
/**
 * Mail Hub Core → Nucleus 2.0 Central Memory Core Connector
 * Built from absolute zero for Surooh Empire
 * 
 * This is a simple Node.js client that Mail Hub can use to connect
 * with Nucleus 2.0 Central Memory Core via WebSocket and REST APIs
 */

const WebSocket = require('ws');
const https = require('https');

class MailHubConnector {
  constructor(config) {
    this.nucleusUrl = config.nucleusUrl || 'https://nucleus.surooh.com';
    this.wsUrl = config.wsUrl || 'wss://nucleus.surooh.com/ws/nucleus';
    this.jwtToken = config.jwtToken; // JWT token for authentication
    this.platform = 'MAIL_HUB';
    
    this.ws = null;
    this.connected = false;
    this.syncInterval = config.syncInterval || 15; // minutes
    this.syncTimer = null;
    
    console.log('[MailHubConnector] Initialized for', this.nucleusUrl);
  }

  /**
   * Connect to Nucleus via WebSocket for real-time communication
   */
  async connectWebSocket() {
    return new Promise((resolve, reject) => {
      this.ws = new WebSocket(this.wsUrl);

      this.ws.on('open', () => {
        console.log('[MailHubConnector] 🔌 WebSocket connected to Nucleus');
        
        // Send authentication
        this.ws.send(JSON.stringify({
          type: 'auth',
          platform: this.platform,
          token: this.jwtToken,
          timestamp: new Date().toISOString()
        }));
      });

      this.ws.on('message', (data) => {
        const msg = JSON.parse(data.toString());
        
        // Resolve promise on successful auth
        if (msg.type === 'ack' && msg.payload.message === 'Authentication successful') {
          resolve();
        }
        
        this.handleMessage(msg);
      });

      this.ws.on('close', () => {
        console.log('[MailHubConnector] WebSocket disconnected');
        this.connected = false;
        
        // Reconnect after 5 seconds
        setTimeout(() => this.connectWebSocket(), 5000);
      });

      this.ws.on('error', (error) => {
        console.error('[MailHubConnector] WebSocket error:', error.message);
        reject(error);
      });
    });
  }

  /**
   * Handle incoming WebSocket message
   */
  handleMessage(message) {
    console.log('[MailHubConnector] 📥 Received:', message.type);

    switch (message.type) {
      case 'ack':
        if (message.payload.message === 'Authentication successful') {
          this.connected = true;
          console.log('[MailHubConnector] ✅ Authenticated as', this.platform);
          
          // Start periodic sync
          this.startPeriodicSync();
        }
        break;

      case 'data':
        // Handle data from Nucleus (AI feedback, models, etc.)
        this.handleNucleusData(message.payload);
        break;

      case 'ping':
        // Respond to ping
        this.ws.send(JSON.stringify({
          type: 'pong',
          timestamp: new Date().toISOString()
        }));
        break;
    }
  }

  /**
   * Handle incoming data from Nucleus
   */
  handleNucleusData(data) {
    console.log('[MailHubConnector] 🧠 AI Feedback from Nucleus:', data);
    
    // Example: Handle AI feedback for email processing
    if (data.dataType === 'AI_FEEDBACK') {
      const { recommendedReply, confidence, modelVersion } = data.payload;
      console.log(`  → Recommended Reply: ${recommendedReply}`);
      console.log(`  → Confidence: ${confidence}`);
      console.log(`  → Model: ${modelVersion}`);
      
      // Your Mail Hub logic here
      // e.g., update local database, trigger auto-reply, etc.
    }
  }

  /**
   * Send email summary to Nucleus via WebSocket
   */
  async sendEmailSummary(emailData) {
    if (!this.connected) {
      console.error('[MailHubConnector] Not connected - queuing message');
      // In production, implement a queue system
      return;
    }

    const message = {
      type: 'data',
      platform: this.platform,
      messageId: `MH-${Date.now()}`,
      payload: {
        platform: this.platform,
        direction: 'OUTBOUND',
        timestamp: new Date().toISOString(),
        dataType: 'EMAIL_SUMMARY',
        payload: {
          messageId: emailData.messageId,
          sender: emailData.sender,
          subject: emailData.subject,
          summary: emailData.summary,
          sentiment: emailData.sentiment,
          category: emailData.category
        },
        metadata: {
          source: 'MAIL_HUB_CORE',
          priority: emailData.priority || 'MEDIUM',
          schemaVersion: 'v1.0.0',
          authToken: this.jwtToken
        }
      },
      timestamp: new Date().toISOString()
    };

    this.ws.send(JSON.stringify(message));
    console.log('[MailHubConnector] 📤 Sent email summary:', emailData.subject);
  }

  /**
   * Send batch sync via REST API (every 15 minutes)
   */
  async sendBatchSync(summaries) {
    try {
      const data = JSON.stringify({
        platform: this.platform,
        direction: 'OUTBOUND',
        timestamp: new Date().toISOString(),
        dataType: 'EMAIL_SUMMARY',
        payload: {
          summary: `Batch sync: ${summaries.length} emails processed`,
          details: summaries
        },
        metadata: {
          source: 'MAIL_HUB_CORE',
          priority: 'MEDIUM',
          schemaVersion: 'v1.0.0',
          authToken: this.jwtToken
        }
      });

      const options = {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': data.length,
          'Authorization': `Bearer ${this.jwtToken}`
        }
      };

      const req = https.request(`${this.nucleusUrl}/api/v1/mailhub/sync`, options, (res) => {
        let responseData = '';
        
        res.on('data', (chunk) => {
          responseData += chunk;
        });

        res.on('end', () => {
          console.log('[MailHubConnector] ✅ Batch sync completed:', responseData);
        });
      });

      req.on('error', (error) => {
        console.error('[MailHubConnector] Batch sync error:', error.message);
      });

      req.write(data);
      req.end();

    } catch (error) {
      console.error('[MailHubConnector] Failed to send batch sync:', error.message);
    }
  }

  /**
   * Start periodic sync (every 15 minutes as per spec)
   */
  startPeriodicSync() {
    if (this.syncTimer) {
      clearInterval(this.syncTimer);
    }

    console.log(`[MailHubConnector] ⏰ Starting periodic sync every ${this.syncInterval} minutes`);

    this.syncTimer = setInterval(() => {
      console.log('[MailHubConnector] 🔄 Periodic sync triggered');
      
      // Collect summaries from Mail Hub local storage
      // const summaries = getLocalSummaries(); // Your implementation
      // this.sendBatchSync(summaries);
      
    }, this.syncInterval * 60 * 1000);
  }

  /**
   * Disconnect from Nucleus
   */
  disconnect() {
    if (this.syncTimer) {
      clearInterval(this.syncTimer);
    }

    if (this.ws) {
      this.ws.close();
    }

    console.log('[MailHubConnector] Disconnected from Nucleus');
  }
}

// ============= Example Usage =============

// Initialize connector
const connector = new MailHubConnector({
  nucleusUrl: 'https://nucleus.surooh.com',
  wsUrl: 'wss://nucleus.surooh.com/ws/nucleus',
  jwtToken: 'YOUR_JWT_TOKEN_HERE', // Get this from Nucleus auth
  syncInterval: 15 // minutes
});

// Connect to Nucleus
connector.connectWebSocket();

// Example: Send email summary when email is processed
/*
connector.sendEmailSummary({
  messageId: 'MH-12345',
  sender: 'client@supplier.com',
  subject: 'New quotation for packaging',
  summary: 'Supplier sent updated quotation for 10k units with delivery timeline.',
  sentiment: 'positive',
  category: 'business',
  priority: 'HIGH'
});
*/

// Graceful shutdown
process.on('SIGINT', () => {
  connector.disconnect();
  process.exit(0);
});

module.exports = MailHubConnector;

File: docs/mailhub-connector.py
"""
Mail Hub Core → Nucleus 2.0 Central Memory Core Connector (Python)
Built from absolute zero for Surooh Empire

This is a simple Python client that Mail Hub can use to connect
with Nucleus 2.0 Central Memory Core via WebSocket and REST APIs
"""

import asyncio
import websockets
import json
import requests
from datetime import datetime
from typing import Dict, Any, Optional

class MailHubConnector:
    def __init__(self, config: Dict[str, Any]):
        self.nucleus_url = config.get('nucleus_url', 'https://nucleus.surooh.com')
        self.ws_url = config.get('ws_url', 'wss://nucleus.surooh.com/ws/nucleus')
        self.jwt_token = config['jwt_token']  # Required: JWT token for authentication
        self.platform = 'MAIL_HUB'
        
        self.ws = None
        self.connected = False
        self.sync_interval = config.get('sync_interval', 15)  # minutes
        
        print(f'[MailHubConnector] Initialized for {self.nucleus_url}')

    async def connect_websocket(self):
        """Connect to Nucleus via WebSocket for real-time communication"""
        try:
            self.ws = await websockets.connect(self.ws_url)
            print('[MailHubConnector] 🔌 WebSocket connected to Nucleus')
            
            # Send authentication
            await self.ws.send(json.dumps({
                'type': 'auth',
                'platform': self.platform,
                'token': self.jwt_token,
                'timestamp': datetime.utcnow().isoformat() + 'Z'
            }))
            
            # Start listening for messages
            await self.listen_messages()
            
        except Exception as e:
            print(f'[MailHubConnector] WebSocket error: {e}')
            # Reconnect after 5 seconds
            await asyncio.sleep(5)
            await self.connect_websocket()

    async def listen_messages(self):
        """Listen for incoming WebSocket messages"""
        try:
            async for message in self.ws:
                data = json.loads(message)
                await self.handle_message(data)
        except websockets.exceptions.ConnectionClosed:
            print('[MailHubConnector] WebSocket disconnected')
            self.connected = False
            # Reconnect
            await asyncio.sleep(5)
            await self.connect_websocket()

    async def handle_message(self, message: Dict[str, Any]):
        """Handle incoming WebSocket message"""
        print(f"[MailHubConnector] 📥 Received: {message.get('type')}")
        
        msg_type = message.get('type')
        
        if msg_type == 'ack':
            if message.get('payload', {}).get('message') == 'Authentication successful':
                self.connected = True
                print(f'[MailHubConnector] ✅ Authenticated as {self.platform}')
                
                # Start periodic sync
                asyncio.create_task(self.start_periodic_sync())
                
        elif msg_type == 'data':
            # Handle data from Nucleus (AI feedback, models, etc.)
            await self.handle_nucleus_data(message.get('payload'))
            
        elif msg_type == 'ping':
            # Respond to ping
            await self.ws.send(json.dumps({
                'type': 'pong',
                'timestamp': datetime.utcnow().isoformat() + 'Z'
            }))

    async def handle_nucleus_data(self, data: Dict[str, Any]):
        """Handle incoming data from Nucleus"""
        print(f'[MailHubConnector] 🧠 AI Feedback from Nucleus: {data}')
        
        # Example: Handle AI feedback for email processing
        if data.get('dataType') == 'AI_FEEDBACK':
            payload = data.get('payload', {})
            recommended_reply = payload.get('recommendedReply')
            confidence = payload.get('confidence')
            model_version = payload.get('modelVersion')
            
            print(f'  → Recommended Reply: {recommended_reply}')
            print(f'  → Confidence: {confidence}')
            print(f'  → Model: {model_version}')
            
            # Your Mail Hub logic here
            # e.g., update local database, trigger auto-reply, etc.

    async def send_email_summary(self, email_data: Dict[str, Any]):
        """Send email summary to Nucleus via WebSocket"""
        if not self.connected:
            print('[MailHubConnector] Not connected - queuing message')
            # In production, implement a queue system
            return
        
        message = {
            'type': 'data',
            'platform': self.platform,
            'messageId': f"MH-{int(datetime.utcnow().timestamp() * 1000)}",
            'payload': {
                'platform': self.platform,
                'direction': 'OUTBOUND',
                'timestamp': datetime.utcnow().isoformat() + 'Z',
                'dataType': 'EMAIL_SUMMARY',
                'payload': {
                    'messageId': email_data['messageId'],
                    'sender': email_data['sender'],
                    'subject': email_data['subject'],
                    'summary': email_data['summary'],
                    'sentiment': email_data.get('sentiment', 'neutral'),
                    'category': email_data.get('category', 'general')
                },
                'metadata': {
                    'source': 'MAIL_HUB_CORE',
                    'priority': email_data.get('priority', 'MEDIUM'),
                    'schemaVersion': 'v1.0.0',
                    'authToken': self.jwt_token
                }
            },
            'timestamp': datetime.utcnow().isoformat() + 'Z'
        }
        
        await self.ws.send(json.dumps(message))
        print(f"[MailHubConnector] 📤 Sent email summary: {email_data['subject']}")

    async def send_batch_sync(self, summaries: list):
        """Send batch sync via REST API (every 15 minutes)"""
        try:
            data = {
                'platform': self.platform,
                'direction': 'OUTBOUND',
                'timestamp': datetime.utcnow().isoformat() + 'Z',
                'dataType': 'EMAIL_SUMMARY',
                'payload': {
                    'summary': f'Batch sync: {len(summaries)} emails processed',
                    'details': summaries
                },
                'metadata': {
                    'source': 'MAIL_HUB_CORE',
                    'priority': 'MEDIUM',
                    'schemaVersion': 'v1.0.0',
                    'authToken': self.jwt_token
                }
            }
            
            headers = {
                'Content-Type': 'application/json',
                'Authorization': f'Bearer {self.jwt_token}'
            }
            
            response = requests.post(
                f'{self.nucleus_url}/api/v1/mailhub/sync',
                json=data,
                headers=headers
            )
            
            print(f'[MailHubConnector] ✅ Batch sync completed: {response.text}')
            
        except Exception as e:
            print(f'[MailHubConnector] Batch sync error: {e}')

    async def start_periodic_sync(self):
        """Start periodic sync (every 15 minutes as per spec)"""
        print(f'[MailHubConnector] ⏰ Starting periodic sync every {self.sync_interval} minutes')
        
        while self.connected:
            await asyncio.sleep(self.sync_interval * 60)  # Convert to seconds
            
            print('[MailHubConnector] 🔄 Periodic sync triggered')
            
            # Collect summaries from Mail Hub local storage
            # summaries = get_local_summaries()  # Your implementation
            # await self.send_batch_sync(summaries)

    async def disconnect(self):
        """Disconnect from Nucleus"""
        if self.ws:
            await self.ws.close()
        
        self.connected = False
        print('[MailHubConnector] Disconnected from Nucleus')


# ============= Example Usage =============

async def main():
    # Initialize connector
    connector = MailHubConnector({
        'nucleus_url': 'https://nucleus.surooh.com',
        'ws_url': 'wss://nucleus.surooh.com/ws/nucleus',
        'jwt_token': 'YOUR_JWT_TOKEN_HERE',  # Get this from Nucleus auth
        'sync_interval': 15  # minutes
    })
    
    # Connect to Nucleus
    await connector.connect_websocket()
    
    # Example: Send email summary when email is processed
    """
    await connector.send_email_summary({
        'messageId': 'MH-12345',
        'sender': 'client@supplier.com',
        'subject': 'New quotation for packaging',
        'summary': 'Supplier sent updated quotation for 10k units with delivery timeline.',
        'sentiment': 'positive',
        'category': 'business',
        'priority': 'HIGH'
    })
    """

if __name__ == '__main__':
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print('[MailHubConnector] Shutting down...')
